pkgs/dart_model/lib/src/json.dart

```diff
@@ -0,0 +1,113 @@
+// Copyright (c) 2024, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:collection/collection.dart';
+
+import 'json_changes.dart';
+
+/// JSON data.
+final class Json {
+  Map<String, Object?> _root;
+
+  Json.fromJson(Map<String, Object?> json) : _root = json;
+
+  JsonChanges computeChangesFrom(Json previous) {
+    final updates = <Update>[];
+    final removals = <Removal>[];
+    _compute(previous._root, _root, Path([]), updates, removals);
+    return JsonChanges(updates: updates, removals: removals);
+  }
+
+  static void _compute(
+      Map<String, Object?> previous,
+      Map<String, Object?> current,
+      Path path,
+      List<Update> updates,
+      List<Removal> removals) {
+    for (final key in previous.keys.followedBy(current.keys).toSet()) {
+      final keyIsInPrevious = previous.containsKey(key);
+      final keyIsInCurrent = current.containsKey(key);
+
+      if (keyIsInPrevious && !keyIsInCurrent) {
+        removals.add(Removal(path: path.followedByOne(key)));
+      } else if (keyIsInPrevious && keyIsInCurrent) {
+        // It's either the same or a change.
+        final previousValue = previous[key]!;
+        final currentValue = current[key]!;
+
+        if (currentValue is Map<String, Object?>) {
+          if (previousValue is Map<String, Object?>) {
+            _compute(previousValue, currentValue, path.followedByOne(key),
+                updates, removals);
+          } else {
+            updates.add(
+                Update(path: path.followedByOne(key), value: currentValue));
+          }
+        } else if (currentValue is String) {
+          if (previousValue is! String || previousValue != currentValue) {
+            updates.add(
+                Update(path: path.followedByOne(key), value: currentValue));
+          }
+        } else if (currentValue is List) {
+          if (previousValue is! List ||
```

Comment:
Don't we probably want incremental updates for lists too?

Reply:
That seems reasonably likely ... see comment elsewhere re: lists of paths vs maps.